File: akka-docs-dev/rst/scala/code/docs/http/HttpServerExampleSpec.scala
/*
 * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
 */

package docs.http

import akka.actor.ActorSystem
import akka.http.scaladsl.model._
import akka.stream.scaladsl._
import akka.stream.testkit.AkkaSpec
import scala.concurrent.Future

class HttpServerExampleSpec
  extends AkkaSpec("akka.actor.default-mailbox.mailbox-type = akka.dispatch.UnboundedMailbox") {
  def ActorSystem(): ActorSystem = system

  "binding example" in {
    //#bind-example
    import akka.http.scaladsl.Http
    import akka.stream.ActorFlowMaterializer

    implicit val system = ActorSystem()
    implicit val materializer = ActorFlowMaterializer()

    val serverSource: Source[Http.IncomingConnection, Future[Http.ServerBinding]] =
      Http(system).bind(interface = "localhost", port = 8080)
    val bindingFuture: Future[Http.ServerBinding] = serverSource.to(Sink.foreach { connection =>
      // foreach materializes the source
      println("Accepted new connection from " + connection.remoteAddress)
      // ... and then actually handle the connection
    }).run()
    //#bind-example
  }

  "full-server-example" in {
    import akka.http.scaladsl.Http
    import akka.stream.ActorFlowMaterializer

    implicit val system = ActorSystem()
    implicit val materializer = ActorFlowMaterializer()

    val serverSource = Http(system).bind(interface = "localhost", port = 8080)

    //#full-server-example
    import akka.http.scaladsl.model.HttpMethods._
    import akka.stream.scaladsl.{ Flow, Sink }

    val requestHandler: HttpRequest => HttpResponse = {
      case HttpRequest(GET, Uri.Path("/"), _, _, _) =>
        HttpResponse(
          entity = HttpEntity(MediaTypes.`text/html`,
            "<html><body>Hello world!</body></html>"))

      case HttpRequest(GET, Uri.Path("/ping"), _, _, _) =>
        HttpResponse(entity = "PONG!")

      case HttpRequest(GET, Uri.Path("/crash"), _, _, _) =>
        sys.error("BOOM!")

      case _: HttpRequest =>
        HttpResponse(404, entity = "Unknown resource!")
    }

    val bindingFuture: Future[Http.ServerBinding] = serverSource.to(Sink.foreach { connection =>
      println("Accepted new connection from " + connection.remoteAddress)

      connection handleWithSyncHandler requestHandler
      // this is equivalent to
      // connection handleWith { Flow[HttpRequest] map requestHandler }
    }).run()
    //#full-server-example
  }
}

File: akka-docs-dev/rst/java/code/docs/http/javadsl/HttpServerExampleSpec.java
/*
 * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
 */

package docs.http.javadsl;

import akka.actor.ActorSystem;
import akka.http.javadsl.Http;
import akka.http.javadsl.IncomingConnection;
import akka.http.javadsl.ServerBinding;
import akka.http.javadsl.model.*;
import akka.http.javadsl.model.ws.Websocket;
import akka.japi.function.Function;
import akka.japi.function.Procedure;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import org.junit.Test;
import scala.Function1;
import scala.concurrent.Await;
import scala.concurrent.Future;
import scala.concurrent.duration.FiniteDuration;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.concurrent.TimeUnit;

public class HttpServerExampleSpec {
    public void bindingExample() {
        //#binding-example
        ActorSystem system = ActorSystem.create();
        Materializer materializer = ActorMaterializer.create(system);

        Source<IncomingConnection, Future<ServerBinding>> serverSource =
            Http.get(system).bind("localhost", 8080, materializer);

        Future<ServerBinding> serverBindingFuture =
            serverSource.to(Sink.foreach(
                new Procedure<IncomingConnection>() {
                    @Override
                    public void apply(IncomingConnection connection) throws Exception {
                        System.out.println("Accepted new connection from " + connection.remoteAddress());
                        // ... and then actually handle the connection
                    }
                })).run(materializer);
        //#binding-example
    }
    public void fullServerExample() {
        //#full-server-example
        ActorSystem system = ActorSystem.create();
        final Materializer materializer = ActorMaterializer.create(system);

        Source<IncomingConnection, Future<ServerBinding>> serverSource =
                Http.get(system).bind("localhost", 8080, materializer);

        final Function<HttpRequest, HttpResponse> requestHandler =
            new Function<HttpRequest, HttpResponse>() {
                private final HttpResponse NOT_FOUND =
                    HttpResponse.create()
                        .withStatus(404)
                        .withEntity("Unknown resource!");

                @Override
                public HttpResponse apply(HttpRequest request) throws Exception {
                    if (request.method() == HttpMethods.GET) {
                        if (request.getUri().path().equals("/"))
                            return
                                HttpResponse.create()
                                    .withEntity(MediaTypes.TEXT_HTML.toContentType(),
                                        "<html><body>Hello world!</body></html>");
                        else if (request.getUri().path().equals("/ping"))
                            return HttpResponse.create().withEntity("PONG!");
                        else
                            return NOT_FOUND;
                    }
                    else return NOT_FOUND;
                }
            };

        Future<ServerBinding> serverBindingFuture =
                serverSource.to(Sink.foreach(
                        new Procedure<IncomingConnection>() {
                            @Override
                            public void apply(IncomingConnection connection) throws Exception {
                                System.out.println("Accepted new connection from " + connection.remoteAddress());

                                connection.handleWithSyncHandler(requestHandler, materializer);
                                // this is equivalent to
                                //connection.handleWith(Flow.of(HttpRequest.class).map(requestHandler), materializer);
                            }
                        })).run(materializer);
        //#full-server-example
    }
}
